### PR TITLE
feat: Segment PDF text into logical blocks (Experience, Education, etc.)

### LINKED ISSUE #6: Segment PDF text into logical blocks (experience, education, etc.)
**Business Value**
Segmenting the extracted text into logical blocks like work experience and education is essential for structuring the résumé data and feeding it into the AI for tailoring. This step improves accuracy and control over the tailored output.

**Description**
Analyze the raw extracted text and segment it into meaningful blocks (e.g., Experience, Education, Skills). Use basic layout heuristics such as headings, bullet points, and line spacing to identify sections.

**Technical Details**
- Implement a parser module `/lib/parser/segmentBlocks.ts`
- Detect section headers like “Experience”, “Education”, “Skills”
- Use keyword anchors and regex to segment blocks
- Handle multi-line bullet points and wrapped descriptions
- Normalize block labels to a standard schema
- Prepare for AI fallback if segmentation fails

**Acceptance Criteria**
- [ ] Can detect at least “Experience” and “Education” blocks reliably
- [ ] Output structure groups text under normalized keys
- [ ] Bullet points and multi-line sections are merged correctly
- [ ] Edge cases (single section or missing header) handled with warnings
- [ ] Unit test validates parsing on at least 2 sample résumé texts

### PR DESCRIPTION
**Pull request by saviobatista**
### 📋 **Overview**
This PR closes #6 and implements a robust PDF text segmentation system that automatically extracts and categorizes résumé content into logical sections (Experience, Education, Skills, etc.). This feature forms the foundation for the AI-powered résumé tailoring system described in the project roadmap.

#### **Core Functionality**
- **`segmentBlocks()` function**: Main segmentation engine that processes extracted PDF text
- **Section Detection**: Intelligent pattern matching for Experience, Education, and Skills sections
- **Content Processing**: Smart merging of bullet points and related content
- **Skills Extraction**: Automatic detection of technical skills from experience and education content
- **Confidence Scoring**: Quality assessment algorithm for segmentation results

#### **New Files Added**
- `lib/parser/segmentBlocks.ts` - Main segmentation implementation (355 lines)
- `lib/parser/config.ts` - Configuration for patterns and constants (90 lines)
- `lib/parser/types.ts` - TypeScript interfaces for internal processing (57 lines)
- `tests/parser/segmentBlocks.test.ts` - Unit tests (188 lines)
- `tests/parser/integration.test.ts` - Integration tests (109 lines)

#### **Enhanced Files**
- `lib/types.ts` - Added public interfaces and error classes (38 lines added)

#### **Type System**
```typescript
interface ResumeBlock {
  content: string;
  confidence: number;
  startLine: number;
  endLine: number;
}

interface SegmentationResult {
  blocks: {
    experience?: string;
    education?: string;
    skills?: string;
    [key: string]: string | undefined;
  };
  warnings: string[];
  metadata: {
    totalBlocks: number;
    confidence: number;
    detectedSections: string[];
  };
}
```

#### **Configuration-Driven**
- **Section Patterns**: Configurable regex patterns for different résumé formats
- **Job Indicators**: Smart detection of job title and company patterns
- **Education Indicators**: Recognition of degree and university patterns
- **Skill Patterns**: Comprehensive technical skills detection

#### **Processing Pipeline**
1. **Text Normalization** → Clean and prepare input text
2. **Section Detection** → Identify section boundaries using patterns
3. **Content Processing** → Merge and structure section content
4. **Skills Extraction** → Extract technical skills from experience/education
5. **Result Building** → Generate structured output with metadata

#### **Test Coverage**
- **16 tests passing** with comprehensive coverage
- **Unit Tests**: Individual function testing with edge cases
- **Integration Tests**: End-to-end pipeline verification
- **Real Data**: Tests use actual résumé sample data

#### **Code Quality**
- ✅ **ESLint**: No warnings or errors
- ✅ **TypeScript**: Compilation successful
- ✅ **Coverage**: All critical paths tested
- ✅ **Error Handling**: Comprehensive error scenarios covered

#### **Smart Content Merging**
```typescript
// Merges bullet points with descriptions
function mergeBulletPoints(content: string): string {
  // Detects job/education patterns and merges related content
  // Handles various bullet point formats (-, •, etc.)
}
```

#### **Confidence Calculation**
```typescript
// Sophisticated scoring based on:
// - Section detection accuracy
// - Content length and quality
// - Pattern matching success
// - Overall structure integrity
```

#### **Error Handling**
```typescript
export class SegmentationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SegmentationError";
  }
}
```

### 📊 **Performance & Scalability**

- **Efficient Processing**: Single-pass text analysis with O(n) complexity
- **Memory Optimized**: Minimal memory footprint during processing
- **Configurable Limits**: Adjustable thresholds for different use cases
- **Extensible**: Easy to add new section types and patterns

#### **With Existing Code**
- **Seamless Integration**: Works with existing `extractText` function
- **Type Compatibility**: Full TypeScript support throughout
- **Error Propagation**: Proper error handling in the pipeline

#### **Future Extensions**
- **AI Integration**: Ready for AI-powered content analysis
- **Template System**: Foundation for LaTeX template generation
- **Custom Patterns**: User-configurable section detection

#### **Input Example**
```
Java Programming Experience
- Senior Java Developer – FinTech Solutions Inc. (2021–2023)
– Led microservices architecture handling $50M monthly transactions
```

#### **Output Structure**
```typescript
{
  blocks: {
    experience: "Senior Java Developer – FinTech Solutions Inc. (2021–2023)\nLed microservices architecture...",
    skills: "Java, microservices, Spring Boot, RESTful APIs, AWS"
  },
  metadata: {
    totalBlocks: 2,
    confidence: 0.85,
    detectedSections: ["experience", "skills"]
  }
}
```

#### **For Users**
- **Automatic Parsing**: No manual résumé section identification needed
- **Structured Data**: Clean, organized content for AI processing
- **Quality Assurance**: Confidence scoring indicates parsing reliability

#### **For Developers**
- **Maintainable Code**: Clean architecture with clear separation of concerns
- **Extensible Design**: Easy to add new features and patterns
- **Comprehensive Testing**: Robust test suite ensures reliability

### 📋 **Testing Instructions**

1. **Run Tests**: `npm test` (all 16 tests should pass)
2. **Lint Check**: `npm run lint` (should be clean)
3. **Type Check**: `npx tsc --noEmit` (should compile successfully)
4. **Integration Test**: Verify PDF extraction + segmentation pipeline

### 🔮 **Future Roadmap Integration**

This feature directly supports the planned roadmap items:
- ✅ **Resume Parser** - Core segmentation functionality
- 🔄 **AI Prompt Generator** - Structured data for AI processing
- 🔄 **Tailored Experience Selector** - Categorized content for customization
-  **LaTeX PDF Renderer** - Structured sections for template generation

### **Breaking Changes**
- **None** - This is a new feature with no impact on existing functionality

### 🧹 **Code Review Checklist**

- [x] **Functionality**: All features work as expected
- [x] **Testing**: Comprehensive test coverage
- [x] **Documentation**: Clear code comments and interfaces
- [x] **Error Handling**: Proper error scenarios covered
- [x] **Performance**: Efficient algorithms and data structures
- [x] **Type Safety**: Full TypeScript support
- [x] **Linting**: Code follows project standards

---

**Ready for Review** 🎯 | **Tests Passing** ✅ | **Production Ready** 🚀

### CODE REVIEW

File: lib/parser/segmentBlocks.ts [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    });
+  } catch (error) {
+    throw new SegmentationError(
+      `Failed to segment text: ${error instanceof Error ? error.message : "Unknown error"}`
+    );
+  }
+}
+
+export { SegmentationError };
+export default segmentBlocks;
```
[the remark this answers is not in the document]
Comment (saviobatista) [reply lib/parser/segmentBlocks.ts:355]: This would raise a problem on typescript because you need to export types and have a default export at same time.

File: tests/parser/integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import { describe, it, expect } from "vitest";
+import { readFile } from "node:fs/promises";
+import { existsSync } from "node:fs";
+import path from "node:path";
+
+import extractText from "../../lib/parser/extractText";
+import segmentBlocks from "../../lib/parser/segmentBlocks";
```
[the remark this answers is not in the document]
Comment (saviobatista) [reply tests/parser/integration.test.ts:7]: Nope, typescript standard.

File: lib/parser/segmentBlocks.ts [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+  if (SECTION_PATTERNS.education.some((pattern) => pattern.test(lowerLabel))) {
+    return "education";
+  }
+  if (SECTION_PATTERNS.skills.some((pattern) => pattern.test(lowerLabel))) {
+    return "skills";
+  }
+
+  return label.toLowerCase().replace(/\s+/g, "_");
+}
+
```
Comment (JoaoMiottiTec) [inline lib/parser/segmentBlocks.ts]: If you use a type to select the keys, I think it's better use a loop over SECTION_PATTERNS.  This way the future maintenance is requires adding a new key  config

File: lib/parser/segmentBlocks.ts [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+      endLine,
+    };
+
+    blocks[normalizedSection].push(sectionContent);
+
+    if (!detectedSections.includes(normalizedSection)) {
+      detectedSections.push(normalizedSection);
+    }
+  }
+}
```
Comment (JoaoMiottiTec) [inline lib/parser/segmentBlocks.ts]: If return early this func plus add a helper, i think it turn the function more legible

File: lib/parser/segmentBlocks.ts [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+/**
+ * Process section content and merge multiple sections of the same type
+ */
+function processSectionContent(blocks: ProcessingBlocks): FinalBlocks {
+  const finalBlocks: FinalBlocks = {};
+
+  for (const [sectionKey, contentArray] of Object.entries(blocks)) {
+    if (sectionKey === "experience" || sectionKey === "education") {
```
Comment (JoaoMiottiTec) [inline lib/parser/segmentBlocks.ts]: I think it's best to define these keys using Set to clear the session because this set prevents duplicate values.

File: lib/parser/segmentBlocks.ts [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    }
+  }
+}
+
+/**
+ * Generate warnings for missing sections and edge cases
+ */
+function generateWarnings(context: WarningContext): string[] {
+  const { finalBlocks, detectedSections } = context;
+  const warnings: string[] = [];
```
Comment (JoaoMiottiTec) [inline lib/parser/segmentBlocks.ts:286]: Set i think it's better again because isn't duplicate if this messages don't send one more time in the  warning

File: tests/parser/integration.test.ts (hunk 2 above)
Comment (JoaoMiottiTec) [inline tests/parser/integration.test.ts]: Alias path i think it's better to future maintenance in the architecture

**Review by JoaoMiottiTec — CHANGES_REQUESTED** [review]

File: lib/parser/segmentBlocks.ts (hunk 6 above)
Comment (saviobatista) [reply lib/parser/segmentBlocks.ts:286]: Not here, this is defined by the business logic and the typed data is only the result holder, as we have specific validation in this method for processing and assigning the warning values, the logic itself avoid duplication of warnings

File: lib/parser/segmentBlocks.ts (hunk 3 above)
Comment (saviobatista) [reply lib/parser/segmentBlocks.ts]: Nice catch, fixed.

File: lib/parser/segmentBlocks.ts (hunk 4 above)
Comment (saviobatista) [reply lib/parser/segmentBlocks.ts]: Refactored

File: tests/parser/integration.test.ts (hunk 2 above)
Comment (saviobatista) [reply tests/parser/integration.test.ts]: Done, also I've refactored all imports from lib and tests folders. It needed also an extra configuration on vite config.

File: lib/parser/segmentBlocks.ts (hunk 5 above)
Comment (saviobatista) [reply lib/parser/segmentBlocks.ts]: Nice catch, refactored.

**Review by JoaoMiottiTec — APPROVED** [review]
